### PR TITLE
Throw a RepositoryConflictException when trying to put or delete a repository in used

### PR DESCRIPTION
**Pull request by gaobinlong**
Relates to #54637.

The main changes are:
1. Add a `RepositoryConflictException` with 409 status code.
2. Throw a `RepositoryConflictException` when trying to put or delete a repository in used.
3. Add unit test for the change.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/snapshots/RepositoriesIT.java [hunk 1]
```diff
@@ -225,4 +231,59 @@ public void testRepositoryVerification() {
...
+
+        String blockedNode = internalCluster().getMasterName();
+        ((MockRepository)internalCluster().getInstance(RepositoriesService.class, blockedNode).repository(repo)).blockOnDataFiles(true);
+        logger.info("--> start deletion of snapshot");
+        ActionFuture<AcknowledgedResponse> future = client().admin().cluster().prepareDeleteSnapshot(repo, snapshot1).execute();
+        logger.info("--> waiting for block to kick in on node [{}]", blockedNode);
+        waitForBlock(blockedNode, repo, TimeValue.timeValueSeconds(10));
+
+        logger.info("--> try deleting the repository, should fail because the deletion of the snapshot is in progress");
+        try {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/snapshots/RepositoriesIT.java:262]: We actually have a neat short-cut for testing this kind of spot with the `expectThrows` method that keeps this much shorter. See https://github.com/elastic/elasticsearch/blob/master/server/src/test/java/org/elasticsearch/snapshots/SharedClusterSnapshotRestoreIT.java#L2169 for an example.

File: server/src/test/java/org/elasticsearch/snapshots/RepositoriesIT.java [hunk 2]
```diff
@@ -225,4 +231,59 @@ public void testRepositoryVerification() {
...
+            client().admin().cluster().prepareDeleteRepository(repo).get();
+            fail("should not be able to delete a repository while it is being used");
+        } catch (RepositoryConflictException e) {
+            assertThat(e.status(),  equalTo(RestStatus.CONFLICT));
+            assertThat(e.getMessage(), containsString("repository conflict," +
+                " trying to modify or unregister repository that is currently used"));
+        }
+
+        logger.info("--> try updating the repository, should fail because the deletion of the snapshot is in progress");
+        try {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/snapshots/RepositoriesIT.java:272]: Same here, `expectThrows` is nicer I think :)

**Review by original-brownbear — CHANGES_REQUESTED** [review]
Thanks @gaobinlong I'm on board with this change in general. I left a few smaller points on the code.

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
@gaobinlong thanks for the changes, looking good. There's a related test failure here though, could you take a look please? (ideally make sure at least `./gradlew :server:check` passes locally) Thanks!

**Comment by gaobinlong** [comment]
@original-brownbear, I have pushed a new commit to fix the test failure  and  `./gradlew :server:check` executed successfully in my local environment.

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by gaobinlong** [comment]
@original-brownbear , all checks have passed, can this PR be approved?

**Comment by fcofdez** [comment]
I can see that there are some merge conflicts and checks failing, could you please merge the latest master branch @gaobinlong? thanks!

**Comment by gaobinlong** [comment]
@fcofdez , I have merged the master branch yet.

**Comment by fcofdez** [comment]
Jenkins test this please

**Comment by gaobinlong** [comment]
@fcofdez, there is a new conflict, I have resolved it and pushed a new commit, can you help to take a look at that?

**Comment by gaobinlong** [comment]
Can we take one step further here?

**Comment by DaveCTurner** [comment]
I think this is a sensible change to make. I resolved the merge conflicts that have accumulated since it was opened and will ping @fcofdez for another review.

**Comment by fcofdez** [comment]
@elasticmachine run tests

**Comment by fcofdez** [comment]
@elasticmachine test this please

**Review by fcofdez — CHANGES_REQUESTED** [review]
Sorry for the delay on this @gaobinlong, the changes look good to me but the test failures seems to be related to the change. We should fix those before merging this patch.

**Comment by gaobinlong** [comment]
@fcofdez I can't see the details of the test failures, because the url shows a 404 error: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request+part-2/5040/ . Can you help to take a look?

**Comment by fcofdez** [comment]
@elasticmachine test this please

**Comment by fcofdez** [comment]
That's strange, I couldn't see the failures either. I've triggered a new CI round to regenerate the results. Sorry about that.

**Comment by fcofdez** [comment]
@gaobinlong this are the failing tests, I think those expect the previous `IllegalArgumentException` or the previous status code.
```
BlobStoreDynamicSettingsIT » testUpdateRateLimitsDynamically FAILED 
SharedClusterSnapshotRestoreIT » testDeleteRepositoryWhileSnapshotting FAILED 
SharedClusterSnapshotRestoreIT » testSnapshotSucceedsAfterSnapshotFailure FAILED
SearchableSnapshotsRepositoryIntegTests » testRepositoryUsedBySearchableSnapshotCanBeUpdatedButNotUnregistered
SearchableSnapshotsClientYamlTestSuiteIT » test {p0=/repository/Delete repository when a snapshot is mounted as an index}
```

**Comment by gaobinlong** [comment]
@fcofdez ,I've fixed the test failures, please help to trigger a new CI, thanks!

**Comment by fcofdez** [comment]
@elasticmachine test this please

**Comment by gaobinlong** [comment]
@fcofdez,  the CI task `:x-pack:plugin:searchable-snapshots:qa:rest:yamlRestTestV7CompatTest` executed failed, we may need backport the change to the 7.x branch, can you help to take a look how to do this?  We should change the `request` to `conflict` here: 
https://github.com/elastic/elasticsearch/blob/a5db81b34c36059b2c1d6d94d434667636f9b0c0/x-pack/plugin/searchable-snapshots/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/repository.yml#L75

**Comment by fcofdez** [comment]
```
diff --git a/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestDeleteRepositoryAction.java b/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestDeleteRepositoryAction.java
index 1198f58b1c0..9506474583a 100644
--- a/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestDeleteRepositoryAction.java
+++ b/server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestDeleteRepositoryAction.java
@@ -8,8 +8,12 @@

 package org.elasticsearch.rest.action.admin.cluster;

+import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.admin.cluster.repositories.delete.DeleteRepositoryRequest;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
 import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.core.RestApiVersion;
 import org.elasticsearch.rest.BaseRestHandler;
 import org.elasticsearch.rest.RestRequest;
 import org.elasticsearch.rest.action.RestToXContentListener;
@@ -40,6 +44,23 @@ public class RestDeleteRepositoryAction extends BaseRestHandler {
         DeleteRepositoryRequest deleteRepositoryRequest = deleteRepositoryRequest(request.param("repository"));
         deleteRepositoryRequest.timeout(request.paramAsTime("timeout", deleteRepositoryRequest.timeout()));
         deleteRepositoryRequest.masterNodeTimeout(request.paramAsTime("master_timeout", deleteRepositoryRequest.masterNodeTimeout()));
-        return channel -> client.admin().cluster().deleteRepository(deleteRepositoryRequest, new RestToXContentListener<>(channel));
+        return channel -> {
+            final RestToXContentListener<ToXContentObject> restListener = new RestToXContentListener<>(channel);
+            client.admin().cluster().deleteRepository(deleteRepositoryRequest, new ActionListener<>() {
+                @Override
+                public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                    restListener.onResponse(acknowledgedResponse);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    if (request.getRestApiVersion().equals(RestApiVersion.V_7)) {
+                        restListener.onFailure(new IllegalStateException(e.getMessage()));
+                    } else {
+                        restListener.onFailure(e);
+                    }
+                }
+            });
+        };
     }
 }
```

@gaobinlong we need to add a little compatibility layer for the `7.x` REST API as we won't backport this change. I'll take care of adding a deprecation notice in `7.x` regarding this change.

**Comment by gaobinlong** [comment]
@fcofdez, all checks have passed yet, can this PR be merged now?

**Comment by fcofdez** [comment]
@gaobinlong sorry for the delay on this. I've pushed a commit handling the BWC dance and docs. Once the CI runs and I check that the docs are fine I'll merge this. Again, sorry for the delay.

**Review by fcofdez — APPROVED** [review]
LGTM

**Comment by fcofdez** [comment]
Thanks for your contribution @gaobinlong !

**Pull request merged**